man/tiff2rgba.1
.\"
.\" Copyright (c) 1988-1997 Sam Leffler
.\" Copyright (c) 1991-1997 Silicon Graphics, Inc.
.\"
.\" Permission to use, copy, modify, distribute, and sell this software and 
.\" its documentation for any purpose is hereby granted without fee, provided
.\" that (i) the above copyright notices and this permission notice appear in
.\" all copies of the software and related documentation, and (ii) the names of
.\" Sam Leffler and Silicon Graphics may not be used in any advertising or
.\" publicity relating to the software without the specific, prior written
.\" permission of Sam Leffler and Silicon Graphics.
.\" 
.\" THE SOFTWARE IS PROVIDED "AS-IS" AND WITHOUT WARRANTY OF ANY KIND, 
.\" EXPRESS, IMPLIED OR OTHERWISE, INCLUDING WITHOUT LIMITATION, ANY 
.\" WARRANTY OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  
.\" 
.\" IN NO EVENT SHALL SAM LEFFLER OR SILICON GRAPHICS BE LIABLE FOR
.\" ANY SPECIAL, INCIDENTAL, INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND,
.\" OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
.\" WHETHER OR NOT ADVISED OF THE POSSIBILITY OF DAMAGE, AND ON ANY THEORY OF 
.\" LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE 
.\" OF THIS SOFTWARE.
.\"
.if n .po 0
.TH TIFF2RGBA 1 "November 2, 2005" "libtiff"
.SH NAME
tiff2rgba \- convert a 
.SM TIFF
image to RGBA color space
.SH SYNOPSIS
.B tiff2rgba
[
.I options
]
.I input.tif
.I output.tif
.SH DESCRIPTION
.I Tiff2rgba
converts a wide variety of TIFF images into an RGBA TIFF image.  This 
includes the ability to translate different color spaces and photometric
interpretation into RGBA, support for alpha blending, and translation
of many different bit depths into a 32bit RGBA image.
.P
Internally this program is implemented using the
.I NDPIReadRGBAImage()
function, and it suffers any limitations of that image.  This includes
limited support for > 8 BitsPerSample images, and flaws with some
esoteric combinations of BitsPerSample, photometric interpretation, 
block organization and planar configuration.  
.P
The generated images are stripped images with four samples per pixel 
(red, green, blue and alpha) or if the
.B \-n
flag is used, three samples
per pixel (red, green, and blue).  The resulting images are always planar
configuration contiguous.  For this reason, this program is a useful utility
for transform exotic TIFF files into a form ingestible by almost any TIFF
supporting software. 
.SH OPTIONS
.TP
.B \-c
Specify a compression scheme to use when writing image data:
.B "\-c none"
for no compression (the default),
.B "\-c packbits"
for the PackBits compression algorithm,
.B "\-c zip"
for the Deflate compression algorithm,
.B "\-c jpeg"
for the JPEG compression algorithm,
and
.B "\-c lzw"
for Lempel-Ziv & Welch.
.TP
.B \-r
Write data with a specified number of rows per strip;
by default the number of rows/strip is selected so that each strip
is approximately 8 kilobytes.
.TP
.B \-b
Process the image one block (strip/tile) at a time instead of by reading
the whole image into memory at once.  This may be necessary for very large
images on systems with limited RAM.
.TP
.B \-n
Drop the alpha component from the output file, producing a pure RGB file.
Currently this does not work if the
.B \-b
flag is also in effect.
.TP
.BI \-M " size"
Set maximum memory allocation size (in MiB). The default is 256MiB.
Set to 0 to disable the limit.
.SH "SEE ALSO"
.BR tiff2bw (1),
.BR NDPIReadRGBAImage (3t),
.BR libtiff (3)
.PP
Libtiff library home page:
.BR http://www.simplesystems.org/libtiff/


ChangeLog
2021-04-16  Even Rouault  <even.rouault@spatialys.com>

	libtiff v4.3.0 released

2021-04-15  Even Rouault  <even.rouault@spatialys.com>

	Merge branch 'libjpeg12_cmake' into 'master'
	libtiff/tif_config.h.cmake.in: surround LIBJPEG_12_PATH by double quotes

	See merge request libtiff/libtiff!237

2021-04-15  Even Rouault  <even.rouault@spatialys.com>

	libtiff/tif_config.h.cmake.in: surround LIBJPEG_12_PATH by double quotes

2021-04-14  Even Rouault  <even.rouault@spatialys.com>

	Merge branch 'lerc_api_change' into 'master'
	tif_lerc.c: cope with API breakage in liblerc master

	See merge request libtiff/libtiff!236

2021-04-14  Even Rouault  <even.rouault@spatialys.com>

	tif_lerc.c: cope with API breakage in liblerc master.

2021-04-08  Even Rouault  <even.rouault@spatialys.com>

	libtiff: remove remaining #ifdef WIN32 and use PRI formatting.

2021-03-10  Even Rouault  <even.rouault@spatialys.com>

	Merge branch 'kmilos-master-patch-73187' into 'master'
	tiffcp: Remove LZW help text, preset not supported

	See merge request libtiff/libtiff!229

2021-03-10  Miloš Komarčević  <miloskomarcevic@aim.com>

	tiffcp: Remove LZW help text, preset not supported.

2021-03-10  Even Rouault  <even.rouault@spatialys.com>

	Merge branch 'feature/lerc' into 'master'
	Add LERC Compression Plugin (closes: #213)

	Closes #213

	See merge request libtiff/libtiff!228

2021-03-10  Even Rouault  <even.rouault@spatialys.com>

	Merge branch 'cmake-fixes' into 'master'
	CMake fixes

	Closes #246 et #245

	See merge request libtiff/libtiff!232

2021-03-09  Roger Leigh  <rleigh@codelibre.net>

	cmake: Correct FindCMath.

	cmake: Correct ZSTD_USABLE typo.

2021-03-07  Roger Leigh  <rleigh@codelibre.net>

	Merge branch 'cmake-find-lib-prefixes' into 'master'
	cmake: Correct find lib prefixes for Deflate and JBIG

	See merge request libtiff/libtiff!231

2021-03-07  Roger Leigh  <rleigh@codelibre.net>

	cmake: FindJBIG uses lib prefix on Windows.

	cmake: FindDeflate uses lib prefix on Windows.

2021-03-07  Even Rouault  <even.rouault@spatialys.com>

	TWebPDecode(): avoid potential overflow on multiplication (CID 1472928)

	NDPIReadDirEntryArrayWithLimit(): avoid false positive from Coverity Scan regarding out-of-bounds access (CID 1472927)

	tif_dirwrite.c: avoid setting a variable that is not read afterwards.

2021-03-07  Even Rouault  <even.rouault@spatialys.com>

	Merge branch 'coverity-fixes' into 'master'
	Coverity fixes (high impact)

	See merge request libtiff/libtiff!227

2021-03-07  Roger Leigh  <rleigh@codelibre.net>

	Fix high-impact Coverity issues (resource leaks).
	The issues are in the tests and tiffcrop, not the core library. Real issues, but not high risk.

	Use to test if Coverity integration is performing properly on merge.

2021-03-07  Even Rouault  <even.rouault@spatialys.com>

	Merge branch 'fix_tif_fax3_encoder_regression' into 'master'
	tif_fax3.c: fix master regression in encoder

	See merge request libtiff/libtiff!230

2021-03-07  Even Rouault  <even.rouault@spatialys.com>

	tif_fax3.c: fix master regression in encoder.
	Fix issue introduced in 39a74eede0455ec8ee334dcddf71f5354d508d8b

	Spotted by gdal's tiff_write_76 test

2021-03-07  Antonio Valentino  <antonio.valentino@tiscali.it>

	Add LERC support in CMake.

2021-03-07  Antonio Valentino  <Antonio.Valentino@esa.int>

	Add LERC support in configure.ac and Makefile.am.

2021-03-07  Antonio Valentino  <antonio.valentino@tiscali.it>

	Add LERC support to tiffcp.

2021-03-07  Antonio Valentino  <Antonio.Valentino@esa.int>

	Add LERC plugin.
	The lerc plugin code has been copyed form GDAL.

2021-03-06  Roger Leigh  <rleigh@codelibre.net>

	Merge branch 'display_tool_purpose' into 'master'
	TIFF tools: insert a line of text summarizing each tool's purpose

	See merge request libtiff/libtiff!214

2021-03-06  Even Rouault  <even.rouault@spatialys.com>

	Merge branch 'cmake-update' into 'master'
	CMake updates

	See merge request libtiff/libtiff!224

2021-03-06  Even Rouault  <even.rouault@spatialys.com>

	tiff.h: typo fix in comment.

2021-02-15  Roger Leigh  <rleigh@codelibre.net>

	Merge branch 'remove-travis' into 'master'
	ci: Remove unused Travis-CI support

	See merge request libtiff/libtiff!226

2021-02-14  Roger Leigh  <rleigh@codelibre.net>

	ci: Remove unused Travis-CI support.

2021-02-14  Roger Leigh  <rleigh@codelibre.net>

	Merge branch 'ci-coverity' into 'master'
	Enable Coverity static analysis with CI pipeline

	See merge request libtiff/libtiff!225

2021-02-14  Roger Leigh  <rleigh@codelibre.net>

	ci: Add Coverity static analysis job.

	ci: Use custom libtiff CI image.

2021-02-13  Roger Leigh  <rleigh@codelibre.net>

	cmake: Add release target.

	cmake: Remove empty contrib files.

	cmake: Tidy toplevel.

	cmake: Move pkg-config support to PkgConfig.cmake.

	cmake: Move library feature options to CXXLibraryFeatures.cmake.

	cmake: Move C++ support to CXXLibrary.cmake.

	cmake: Add FindCMath to handle libm linking portably.

	cmake: Tidy unused includes.

	cmake: Rename release date to build date.

	cmake: Compute timestamp portably.

	cmake: Remove remaining uses of report_values()

	cmake: Move JPEG12 checks to JPEGCodec.cmake.

	cmake: Move OpenGL checks to OpenGLChecks.cmake.

	cmake: Move OpenGL checks to OpenGLChecks.cmake.

	cmake: Move WebP codec support to WebPCodec.cmake.

	cmake: Use imported targets for WebP.

	cmake: Add FindWebP.

	cmake: Move ZSTD codec support to ZSTDCodec.cmake.

	cmake: Use imported targets for ZSTD.

	cmake: Add FindZSTD.

	cmake: Move LZMA codec support to LZMACodec.cmake.

	cmake: Use imported targets for LibLZMA.

	cmake: Move JBIG codec support to JBIGCodec.cmake.

	cmake: Use imported targets for JBIG.

	cmake: Add FindJBIG.

	cmake: Move PixarLog codec support to PixarLogCodec.cmake.

	cmake: Report system name in configuration report.

	cmake: Move JPEG codec support to JPEGCodec.cmake.

	cmake: Use imported targets for JPEG.


tools/tiff2rgba.c
/*
 * Copyright (c) 1991-1997 Sam Leffler
 * Copyright (c) 1991-1997 Silicon Graphics, Inc.
 *
 * Permission to use, copy, modify, distribute, and sell this software and 
 * its documentation for any purpose is hereby granted without fee, provided
 * that (i) the above copyright notices and this permission notice appear in
 * all copies of the software and related documentation, and (ii) the names of
 * Sam Leffler and Silicon Graphics may not be used in any advertising or
 * publicity relating to the software without the specific, prior written
 * permission of Sam Leffler and Silicon Graphics.
 *
 * THE SOFTWARE IS PROVIDED "AS-IS" AND WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS, IMPLIED OR OTHERWISE, INCLUDING WITHOUT LIMITATION, ANY
 * WARRANTY OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
 *
 * IN NO EVENT SHALL SAM LEFFLER OR SILICON GRAPHICS BE LIABLE FOR
 * ANY SPECIAL, INCIDENTAL, INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND,
 * OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
 * WHETHER OR NOT ADVISED OF THE POSSIBILITY OF DAMAGE, AND ON ANY THEORY OF
 * LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE
 * OF THIS SOFTWARE.
 */

#include "tif_config.h"
#include "libport.h"

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif

#include "tiffiop.h"
#include "tiffio.h"

#ifndef EXIT_SUCCESS
#define EXIT_SUCCESS 0
#endif
#ifndef EXIT_FAILURE
#define EXIT_FAILURE 1
#endif

#define	streq(a,b)	(strcmp(a,b) == 0)
#define	CopyField(tag, v) \
    if (NDPIGetField(in, tag, &v)) NDPISetField(out, tag, v)

#ifndef howmany
#define	howmany(x, y)	(((x)+((y)-1))/(y))
#endif
#define	roundup(x, y)	(howmany(x,y)*((uint32_t)(y)))

static uint16_t compression = COMPRESSION_PACKBITS;
static uint32_t rowsperstrip = (uint32_t) -1;
static int process_by_block = 0; /* default is whole image at once */
static int no_alpha = 0;
static int bigtiff_output = 0;
#define DEFAULT_MAX_MALLOC (256 * 1024 * 1024)
/* malloc size limit (in bytes)
 * disabled when set to 0 */
static tmsize_t maxMalloc = DEFAULT_MAX_MALLOC;


static int tiffcvt(TIFF* in, TIFF* out);
static void usage(int code);

int
main(int argc, char* argv[])
{
	TIFF *in, *out;
	int c;
#if !HAVE_DECL_OPTARG
	extern int optind;
	extern char *optarg;
#endif

	while ((c = getopt(argc, argv, "c:r:t:bn8hM:")) != -1)
		switch (c) {
			case 'M':
				maxMalloc = (tmsize_t)strtoul(optarg, NULL, 0) << 20;
				break;
			case 'b':
				process_by_block = 1;
				break;

			case 'c':
				if (streq(optarg, "none"))
					compression = COMPRESSION_NONE;
				else if (streq(optarg, "packbits"))
					compression = COMPRESSION_PACKBITS;
				else if (streq(optarg, "lzw"))
					compression = COMPRESSION_LZW;
				else if (streq(optarg, "jpeg"))
					compression = COMPRESSION_JPEG;
				else if (streq(optarg, "zip"))
					compression = COMPRESSION_DEFLATE;
				else
					usage(EXIT_FAILURE);
				break;

			case 'r':
				rowsperstrip = atoi(optarg);
				break;

			case 't':
				rowsperstrip = atoi(optarg);
				break;

			case 'n':
				no_alpha = 1;
				break;

			case '8':
				bigtiff_output = 1;
				break;

			case 'h':
				usage(EXIT_SUCCESS);
				/*NOTREACHED*/
				break;
			case '?':
				usage(EXIT_FAILURE);
				/*NOTREACHED*/
				break;
		}

	if (argc - optind < 2)
		usage(EXIT_FAILURE);

	out = NDPIOpen(argv[argc-1], bigtiff_output?"w8":"w");
	if (out == NULL)
		return (EXIT_FAILURE);

	for (; optind < argc-1; optind++) {
		in = NDPIOpen(argv[optind], "r");
		if (in != NULL) {
			do {
				if (!tiffcvt(in, out) ||
				    !NDPIWriteDirectory(out)) {
					(void) NDPIClose(out);
					(void) NDPIClose(in);
					return (1);
				}
			} while (NDPIReadDirectory(in));
			(void) NDPIClose(in);
		}
	}
	(void) NDPIClose(out);
	return (EXIT_SUCCESS);
}

static int
cvt_by_tile( TIFF *in, TIFF *out )

{
    uint32_t* raster;			/* retrieve RGBA image */
    uint32_t  width, height;		/* image width & height */
    uint32_t  tile_width, tile_height;
    uint32_t  row, col;
    uint32_t  *wrk_line;
    int	    ok = 1;
    uint32_t  rastersize, wrk_linesize;

    NDPIGetField(in, TIFFTAG_IMAGEWIDTH, &width);
    NDPIGetField(in, TIFFTAG_IMAGELENGTH, &height);

    if( !NDPIGetField(in, TIFFTAG_TILEWIDTH, &tile_width)
        || !NDPIGetField(in, TIFFTAG_TILELENGTH, &tile_height) ) {
        NDPIError(NDPIFileName(in), "Source image not tiled");
        return (0);
    }
    
    NDPISetField(out, TIFFTAG_TILEWIDTH, tile_width );
    NDPISetField(out, TIFFTAG_TILELENGTH, tile_height );

    /*
     * Allocate tile buffer
     */
    rastersize = tile_width * tile_height * sizeof (uint32_t);
    if (tile_width != (rastersize / tile_height) / sizeof( uint32_t))
    {
	NDPIError(NDPIFileName(in), "Integer overflow when calculating raster buffer");
	exit(EXIT_FAILURE);
    }
    raster = (uint32_t*)_NDPImalloc(rastersize);
    if (raster == 0) {
        NDPIError(NDPIFileName(in), "No space for raster buffer");
        return (0);
    }

    /*
     * Allocate a scanline buffer for swapping during the vertical
     * mirroring pass.
     */
    wrk_linesize = tile_width * sizeof (uint32_t);
    if (tile_width != wrk_linesize / sizeof (uint32_t))
    {
        NDPIError(NDPIFileName(in), "Integer overflow when calculating wrk_line buffer");
	exit(EXIT_FAILURE);
    }
    wrk_line = (uint32_t*)_NDPImalloc(wrk_linesize);
    if (!wrk_line) {
        NDPIError(NDPIFileName(in), "No space for raster scanline buffer");
        ok = 0;
    }
    
    /*
     * Loop over the tiles.
     */
    for( row = 0; ok && row < height; row += tile_height )
    {
        for( col = 0; ok && col < width; col += tile_width )
        {
            uint32_t i_row;

            /* Read the tile into an RGBA array */
            if (!NDPIReadRGBATile(in, col, row, raster)) {
                ok = 0;
                break;
            }


	    /*
	     * XXX: raster array has 4-byte unsigned integer type, that is why
	     * we should rearrange it here.
	     */
#if HOST_BIGENDIAN
	    NDPISwabArrayOfLong(raster, tile_width * tile_height);
#endif

            /*
             * For some reason the NDPIReadRGBATile() function chooses the
             * lower left corner as the origin.  Vertically mirror scanlines.
             */
            for( i_row = 0; i_row < tile_height / 2; i_row++ )
            {
                uint32_t	*top_line, *bottom_line;

                top_line = raster + tile_width * i_row;
                bottom_line = raster + tile_width * (tile_height-i_row-1);

                _NDPImemcpy(wrk_line, top_line, 4*tile_width);
                _NDPImemcpy(top_line, bottom_line, 4*tile_width);
                _NDPImemcpy(bottom_line, wrk_line, 4*tile_width);
            }

            /*
             * Write out the result in a tile.
             */

            if( NDPIWriteEncodedTile( out,
                                      NDPIComputeTile( out, col, row, 0, 0),
                                      raster,
                                      4 * tile_width * tile_height ) == -1 )
            {
                ok = 0;
                break;
            }
        }
    }

    _NDPIfree( raster );
    _NDPIfree( wrk_line );

    return ok;
}

static int
cvt_by_strip( TIFF *in, TIFF *out )

{
    uint32_t* raster;			/* retrieve RGBA image */
    uint32_t  width, height;		/* image width & height */
    uint32_t  row;
    uint32_t  *wrk_line;
    int	    ok = 1;
    uint32_t  rastersize, wrk_linesize;

    NDPIGetField(in, TIFFTAG_IMAGEWIDTH, &width);
    NDPIGetField(in, TIFFTAG_IMAGELENGTH, &height);

    if( !NDPIGetField(in, TIFFTAG_ROWSPERSTRIP, &rowsperstrip) ) {
        NDPIError(NDPIFileName(in), "Source image not in strips");
        return (0);
    }
    
    NDPISetField(out, TIFFTAG_ROWSPERSTRIP, rowsperstrip);

    /*
     * Allocate strip buffer
     */
    rastersize = width * rowsperstrip * sizeof (uint32_t);
    if (width != (rastersize / rowsperstrip) / sizeof( uint32_t))
    {
	NDPIError(NDPIFileName(in), "Integer overflow when calculating raster buffer");
	exit(EXIT_FAILURE);
    }
    raster = (uint32_t*)_NDPImalloc(rastersize);
    if (raster == 0) {
        NDPIError(NDPIFileName(in), "No space for raster buffer");
        return (0);
    }

    /*
     * Allocate a scanline buffer for swapping during the vertical
     * mirroring pass.
     */
    wrk_linesize = width * sizeof (uint32_t);
    if (width != wrk_linesize / sizeof (uint32_t))
    {
        NDPIError(NDPIFileName(in), "Integer overflow when calculating wrk_line buffer");
	exit(EXIT_FAILURE);
    }
    wrk_line = (uint32_t*)_NDPImalloc(wrk_linesize);
    if (!wrk_line) {
        NDPIError(NDPIFileName(in), "No space for raster scanline buffer");
        ok = 0;
    }
    
    /*
     * Loop over the strips.
     */
    for( row = 0; ok && row < height; row += rowsperstrip )
    {
        int	rows_to_write, i_row;

        /* Read the strip into an RGBA array */
        if (!NDPIReadRGBAStrip(in, row, raster)) {
            ok = 0;
            break;
        }

	/*
	 * XXX: raster array has 4-byte unsigned integer type, that is why
	 * we should rearrange it here.
	 */
#if HOST_BIGENDIAN
	NDPISwabArrayOfLong(raster, width * rowsperstrip);
#endif

        /*
         * Figure out the number of scanlines actually in this strip.
         */
        if( row + rowsperstrip > height )
            rows_to_write = height - row;
        else
            rows_to_write = rowsperstrip;

        /*
         * For some reason the NDPIReadRGBAStrip() function chooses the
         * lower left corner as the origin.  Vertically mirror scanlines.
         */

        for( i_row = 0; i_row < rows_to_write / 2; i_row++ )
        {
            uint32_t	*top_line, *bottom_line;

            top_line = raster + width * i_row;
            bottom_line = raster + width * (rows_to_write-i_row-1);

            _NDPImemcpy(wrk_line, top_line, 4*width);
            _NDPImemcpy(top_line, bottom_line, 4*width);
            _NDPImemcpy(bottom_line, wrk_line, 4*width);
        }

        /*
         * Write out the result in a strip
         */

        if( NDPIWriteEncodedStrip( out, row / rowsperstrip, raster,
                                   4 * rows_to_write * width ) == -1 )
        {
            ok = 0;
            break;
        }
    }

    _NDPIfree( raster );
    _NDPIfree( wrk_line );

    return ok;
}

/*
 * cvt_whole_image()
 *
 * read the whole image into one big RGBA buffer and then write out
 * strips from that.  This is using the traditional NDPIReadRGBAImage()
 * API that we trust.
 */

static int
cvt_whole_image( TIFF *in, TIFF *out )

{
    uint32_t* raster;			/* retrieve RGBA image */
    uint32_t  width, height;		/* image width & height */
    uint32_t  row;
    size_t pixel_count;
        
    NDPIGetField(in, TIFFTAG_IMAGEWIDTH, &width);
    NDPIGetField(in, TIFFTAG_IMAGELENGTH, &height);
    pixel_count = width * height;

    /* XXX: Check the integer overflow. */
    if (!width || !height || pixel_count / width != height) {
        NDPIError(NDPIFileName(in),
		  "Malformed input file; can't allocate buffer for raster of %"PRIu32"x%"PRIu32" size",
		  width, height);
        return 0;
    }
    if (maxMalloc != 0 && (tmsize_t)pixel_count * (tmsize_t)sizeof(uint32_t) > maxMalloc) {
	NDPIError(NDPIFileName(in),
		  "Raster size %"TIFF_SIZE_FORMAT" over memory limit (%" TIFF_SSIZE_FORMAT "), try -b option.",
              pixel_count * sizeof(uint32_t), maxMalloc);
        return 0;
    }

    rowsperstrip = NDPIDefaultStripSize(out, rowsperstrip);
    NDPISetField(out, TIFFTAG_ROWSPERSTRIP, rowsperstrip);

    raster = (uint32_t*)_NDPICheckMalloc(in, pixel_count, sizeof(uint32_t), "raster buffer");
    if (raster == 0) {
        NDPIError(NDPIFileName(in), "Failed to allocate buffer (%"TIFF_SIZE_FORMAT" elements of %"TIFF_SIZE_FORMAT" each)",
		  pixel_count, sizeof(uint32_t));
        return (0);
    }

    /* Read the image in one chunk into an RGBA array */
    if (!NDPIReadRGBAImageOriented(in, width, height, raster,
                                   ORIENTATION_TOPLEFT, 0)) {
        _NDPIfree(raster);
        return (0);
    }

    /*
     * XXX: raster array has 4-byte unsigned integer type, that is why
     * we should rearrange it here.
     */
#if HOST_BIGENDIAN
    NDPISwabArrayOfLong(raster, width * height);
#endif

    /*
     * Do we want to strip away alpha components?
     */
    if (no_alpha)
    {
        size_t count = pixel_count;
        unsigned char *src, *dst;

	src = dst = (unsigned char *) raster;
        while (count > 0)
        {
	    *(dst++) = *(src++);
	    *(dst++) = *(src++);
	    *(dst++) = *(src++);
	    src++;
	    count--;
        }
    }

    /*
     * Write out the result in strips
     */
    for (row = 0; row < height; row += rowsperstrip)
    {
        unsigned char * raster_strip;
        int	rows_to_write;
        int	bytes_per_pixel;

        if (no_alpha)
        {
            raster_strip = ((unsigned char *) raster) + 3 * row * width;
            bytes_per_pixel = 3;
        }
        else
        {
            raster_strip = (unsigned char *) (raster + row * width);
            bytes_per_pixel = 4;
        }

        if( row + rowsperstrip > height )
            rows_to_write = height - row;
        else
            rows_to_write = rowsperstrip;

        if( NDPIWriteEncodedStrip( out, row / rowsperstrip, raster_strip,
                             bytes_per_pixel * rows_to_write * width ) == -1 )
        {
            _NDPIfree( raster );
            return 0;
        }
    }

    _NDPIfree( raster );

    return 1;
}


static int
tiffcvt(TIFF* in, TIFF* out)
{
	uint32_t width, height;		/* image width & height */
	uint16_t shortv;
	float floatv;
	char *stringv;
	uint32_t longv;
        uint16_t v[1];

	NDPIGetField(in, TIFFTAG_IMAGEWIDTH, &width);
	NDPIGetField(in, TIFFTAG_IMAGELENGTH, &height);

	CopyField(TIFFTAG_SUBFILETYPE, longv);
	NDPISetField(out, TIFFTAG_IMAGEWIDTH, width);
	NDPISetField(out, TIFFTAG_IMAGELENGTH, height);
	NDPISetField(out, TIFFTAG_BITSPERSAMPLE, 8);
	NDPISetField(out, TIFFTAG_COMPRESSION, compression);
	NDPISetField(out, TIFFTAG_PHOTOMETRIC, PHOTOMETRIC_RGB);

	CopyField(TIFFTAG_FILLORDER, shortv);
	NDPISetField(out, TIFFTAG_ORIENTATION, ORIENTATION_TOPLEFT);

        if( no_alpha )
            NDPISetField(out, TIFFTAG_SAMPLESPERPIXEL, 3);
        else
            NDPISetField(out, TIFFTAG_SAMPLESPERPIXEL, 4);

        if( !no_alpha )
        {
            v[0] = EXTRASAMPLE_ASSOCALPHA;
            NDPISetField(out, TIFFTAG_EXTRASAMPLES, 1, v);
        }

	CopyField(TIFFTAG_XRESOLUTION, floatv);
	CopyField(TIFFTAG_YRESOLUTION, floatv);
	CopyField(TIFFTAG_RESOLUTIONUNIT, shortv);
	NDPISetField(out, TIFFTAG_PLANARCONFIG, PLANARCONFIG_CONTIG);
	NDPISetField(out, TIFFTAG_SOFTWARE, TIFFGetVersion());
	CopyField(TIFFTAG_DOCUMENTNAME, stringv);

	if (maxMalloc != 0 && NDPIStripSize(in) > maxMalloc)
	{
		NDPIError(NDPIFileName(in),
			  "Strip Size %" TIFF_SSIZE_FORMAT " over memory limit (%" TIFF_SSIZE_FORMAT ")",
                  NDPIStripSize(in), maxMalloc);
		return 0;
	}
        if( process_by_block && NDPIIsTiled( in ) )
            return( cvt_by_tile( in, out ) );
        else if( process_by_block )
            return( cvt_by_strip( in, out ) );
        else
            return( cvt_whole_image( in, out ) );
}

static const char usage_info[] =
/* Help information format modified for the sake of consistency with the other tiff tools */
/*    "usage: tiff2rgba [-c comp] [-r rows] [-b] [-n] [-8] [-M size] input... output" */
/*     "where comp is one of the following compression algorithms:" */
"Convert a TIFF image to RGBA color space\n\n"
"usage: tiff2rgba [options] input output\n"
"where options are:\n"
#ifdef JPEG_SUPPORT
" -c jpeg      JPEG encoding\n"
#endif
#ifdef ZIP_SUPPORT
" -c zip       Zip/Deflate encoding\n"
#endif
#ifdef LZW_SUPPORT
" -c lzw       Lempel-Ziv & Welch encoding\n"
#endif
#ifdef PACKBITS_SUPPORT
" -c packbits  PackBits encoding\n"
#endif
#if defined(JPEG_SUPPORT) || defined(ZIP_SUPPORT) || defined(LZW_SUPPORT) || defined(PACKBITS_SUPPORT)
" -c none      no compression\n"
#endif
"\n"
/* "and the other options are:\n" */
" -r rows/strip\n"
" -b (progress by block rather than as a whole image)\n"
" -n don't emit alpha component.\n"
" -8 write BigTIFF file instead of ClassicTIFF\n"
" -M set the memory allocation limit in MiB. 0 to disable limit\n"
;

static void
usage(int code)
{
	FILE * out = (code == EXIT_SUCCESS) ? stdout : stderr;

        fprintf(out, "%s\n\n", TIFFGetVersion());
        fprintf(out, "%s", usage_info);
	exit(code);
}

/* vim: set ts=8 sts=8 sw=8 noet: */
/*
 * Local Variables:
 * mode: c
 * c-basic-offset: 8
 * fill-column: 78
 * End:
 */
